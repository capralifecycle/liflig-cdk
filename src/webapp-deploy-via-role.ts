import {
  WebappDeploy,
  type WebappDeployProps,
} from "@capraconsulting/webapp-deploy-lambda"
import * as cdk from "aws-cdk-lib"
import * as iam from "aws-cdk-lib/aws-iam"
import * as constructs from "constructs"

interface Props {
  /**
   * Reference to the IAM Role that will be granted permission to
   * assume the CI role. This role must have permission to assume
   * the CI role.
   */
  externalRoleArn: string
  /**
   * The name of the role that will be created.
   */
  roleName: string
  webappDeploy: WebappDeployProps
}

/**
 * Provide the construct described at
 * https://github.com/capraconsulting/webapp-deploy-lambda
 * in combination with a separate role to be used to
 * trigger the process.
 */
export class WebappDeployViaRole extends constructs.Construct {
  constructor(scope: constructs.Construct, id: string, props: Props) {
    super(scope, id)

    const roleToBeAssumedForDeploy = new iam.Role(this, "Role", {
      assumedBy: new iam.ArnPrincipal(props.externalRoleArn),
      roleName: props.roleName,
    })

    const webappDeploy = new WebappDeploy(this, "Resource", props.webappDeploy)

    webappDeploy.deployFn.grantInvoke(roleToBeAssumedForDeploy)

    new cdk.CfnOutput(this, "FunctionArnOutput", {
      value: webappDeploy.deployFn.functionArn,
    })
    new cdk.CfnOutput(this, "RoleArnOutput", {
      value: roleToBeAssumedForDeploy.roleArn,
    })
  }
}
