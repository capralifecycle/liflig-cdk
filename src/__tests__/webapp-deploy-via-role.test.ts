import { Distribution } from "aws-cdk-lib/aws-cloudfront"
import { Bucket } from "aws-cdk-lib/aws-s3"
import { App, Stack } from "aws-cdk-lib"
import "jest-cdk-snapshot"
import { WebappDeployViaRole } from "../webapp-deploy-via-role"

test("webapp-deploy-via-role", () => {
  const app = new App()
  const stack1 = new Stack(app, "Stack1")

  const buildsBucket = Bucket.fromBucketName(
    stack1,
    "BuildsBucket",
    "bucket-name",
  )

  const webBucket = new Bucket(stack1, "WebBucket")

  const distribution = Distribution.fromDistributionAttributes(
    stack1,
    "Distribution",
    {
      distributionId: "EKJ2IPY1KTEAR1",
      domainName: "example.com",
    },
  )

  new WebappDeployViaRole(stack1, "WebappDeploy", {
    externalRoleArn: "arn:aws:iam::112233445566:role/some-role",
    roleName: "my-role",
    webappDeploy: {
      buildsBucket,
      webBucket,
      distribution,
    },
  })

  expect(stack1).toMatchCdkSnapshot()
})
