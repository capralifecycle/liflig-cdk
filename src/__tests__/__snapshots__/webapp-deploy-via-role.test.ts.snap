// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`webapp-deploy-via-role 1`] = `
Object {
  "Outputs": Object {
    "WebappDeployFunctionArnOutputDA242124": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "WebappDeployWebappDeployResourceD148398D",
          "Arn",
        ],
      },
    },
    "WebappDeployRoleArnOutputF50737C4": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "WebappDeployRole6BC5D130",
          "Arn",
        ],
      },
    },
  },
  "Resources": Object {
    "WebBucket12880F5B": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "WebappDeployProviderframeworkonEvent93982F31": Object {
      "DependsOn": Array [
        "WebappDeployProviderframeworkonEventServiceRoleDefaultPolicy81F824E2",
        "WebappDeployProviderframeworkonEventServiceRoleF0B4B311",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "bdc104ed9cab1b5b6421713c8155f0b753380595356f710400609664d3635eca.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (Stack1/WebappDeploy/Resource/Provider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "WebappDeployWebappDeployResourceD148398D",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "LoggingConfig": Object {
          "ApplicationLogLevel": "FATAL",
          "LogFormat": "JSON",
        },
        "Role": Object {
          "Fn::GetAtt": Array [
            "WebappDeployProviderframeworkonEventServiceRoleF0B4B311",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebappDeployProviderframeworkonEventServiceRoleDefaultPolicy81F824E2": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "WebappDeployWebappDeployResourceD148398D",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "WebappDeployWebappDeployResourceD148398D",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "lambda:GetFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "WebappDeployWebappDeployResourceD148398D",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WebappDeployProviderframeworkonEventServiceRoleDefaultPolicy81F824E2",
        "Roles": Array [
          Object {
            "Ref": "WebappDeployProviderframeworkonEventServiceRoleF0B4B311",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WebappDeployProviderframeworkonEventServiceRoleF0B4B311": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WebappDeployRole6BC5D130": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "arn:aws:iam::112233445566:role/some-role",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "my-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "WebappDeployRoleDefaultPolicy4A1D7118": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "WebappDeployWebappDeployResourceD148398D",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "WebappDeployWebappDeployResourceD148398D",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WebappDeployRoleDefaultPolicy4A1D7118",
        "Roles": Array [
          Object {
            "Ref": "WebappDeployRole6BC5D130",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WebappDeployWebappDeployResourceD148398D": Object {
      "DependsOn": Array [
        "WebappDeployWebappDeployResourceServiceRoleDefaultPolicyBD628B6A",
        "WebappDeployWebappDeployResourceServiceRoleE152E052",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "04d04d3e5f2133139b32c63f4c1a89b8f505d7fe7917cccadbda1c02098f0f45.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "CF_DISTRIBUTION_ID": "EKJ2IPY1KTEAR1",
            "DEPLOY_LOG_BUCKET_URL": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "s3://",
                  Object {
                    "Ref": "WebBucket12880F5B",
                  },
                  "/deployments.log",
                ],
              ],
            },
            "EXPIRE_SECONDS": "432000",
            "TARGET_BUCKET_URL": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "s3://",
                  Object {
                    "Ref": "WebBucket12880F5B",
                  },
                  "/web",
                ],
              ],
            },
          },
        },
        "Handler": "webapp_deploy.main.handler",
        "ReservedConcurrentExecutions": 1,
        "Role": Object {
          "Fn::GetAtt": Array [
            "WebappDeployWebappDeployResourceServiceRoleE152E052",
            "Arn",
          ],
        },
        "Runtime": "python3.12",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebappDeployWebappDeployResourceServiceRoleDefaultPolicyBD628B6A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudfront:CreateInvalidation",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "WebBucket12880F5B",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "WebBucket12880F5B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::bucket-name",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::bucket-name/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WebappDeployWebappDeployResourceServiceRoleDefaultPolicyBD628B6A",
        "Roles": Array [
          Object {
            "Ref": "WebappDeployWebappDeployResourceServiceRoleE152E052",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WebappDeployWebappDeployResourceServiceRoleE152E052": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
